### PR TITLE
SWE6 Take Attendance

### LINKED ISSUE #6: Instructors Take Attendance
As an Instructor, I need to click on seats where students are absent on each lecture day, in order to record who is absent.

Date selection should default to the current date.

### LINKED ISSUE #75: Take Attendance Backend
## Issue Description
This issue involves the creation of the backend for taking attendance.

## Tasks
- [x] Create absence entity & add lists in user and course
- [x] Controller method for REST endpoint

### LINKED ISSUE #76: Take Attendance Frontend
## Issue Description
This issue involves the frontend logic for taking attendance.

## Tasks
- [x] Create attendance, submit and cancel buttons
- [x] Create click logic for grid
- [x] Create the ajax post request method
- [x] Create toaster for success / failure
- [x] Create UI indicator for "absent"

edit: removing json template

### PR DESCRIPTION
**Pull request by jsmith**
## Taking Attendance
fixes #6 
fixes #75 
fixes #76 

### Changes Proposed in this Pull Request:
- Frontend logic so that instructors can take attendance
- Backend controller method to process the absent students
- A few other small issues (index button size, BadRequestException code)

### CODE REVIEW

File: src/main/java/com/unb/tracker/web/AbsenceController.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        List<Absence> absences = absenceRepository.findByCourseIdAndDate(courseId, today);
+
+        return absences;
+    }
+
+    @GetMapping(value = "/courses/{courseId}/absences/{student}")
+    public List<Absence> getAllStudentAbsencesById(@PathVariable String courseId, @PathVariable String studentId) {
+        LOG.info("getAllStudentAbsences - starting - courseId: {}, student: {}", courseId, studentId);
+
+        java.sql.Date today = new java.sql.Date(new java.util.Date().getTime());
```
Comment (tonerm43) [inline src/main/java/com/unb/tracker/web/AbsenceController.java]: Do we need this?

Comment (jsmith) [reply src/main/java/com/unb/tracker/web/AbsenceController.java]: No we don't need `today`, I'm not sure if these methods are used actually!

**Comment by brandon1024** [comment]
These changes look good to me :) @jacsmith21 can you fix the merge conflicts?

**Comment by jsmith** [comment]
Sure I'll fix those tonight!

**Comment by mwalz1** [comment]
@jacsmith21 Should we make an absence entry unique through its: date, course_id & student_id?  Re-submitting attendance creates a lot of duplicate entries in our database.

**Comment by jsmith** [comment]
@mwalz1 I'm not sure, is that a strict requirement from Dr. MacIsaac? It wouldn't be too hard to implement

**Comment by jsmith** [comment]
Hey, so there are a few more requirements being implemented for tomorrow:

* When an instructor submits attendance more than one time per day, it will replace whatever is in the database
* After an instructor submits their attendance, they should be able to see the attendance statuses until the following day

**Review by mwalz1 — CHANGES_REQUESTED** [review]
Implement the following requirements:

 * When an instructor submits attendance more than one time per day, it will replace whatever is in the database

 * After an instructor submits their attendance, they should be able to see the attendance statuses until the following day

**Comment by jsmith** [comment]
I will be pushing changes within the next hour or two!

**Review by mwalz1 — APPROVED** [review]
Looks good.  Great job delivering changes after the additional requirements arrived.

File: src/main/java/com/unb/tracker/web/CourseController.java [hunk 2]
```diff
@@ -185,15 +187,31 @@ public String courseSave(@ModelAttribute Course course, ModelMap map, Principal
...
             course.setSection("");
-            courseRepository.save(course);
-
-            map.addAttribute("course", course);
-            return "redirect:/" + user.getUsername() + "/" + course.getName();
         }
 
         courseRepository.save(course);
+        
+        if(newCourse) {
```
Comment (jsmith) [inline src/main/java/com/unb/tracker/web/CourseController.java:194]: This is a weird fix to the issue we were having but it works. 

Basically, the seats weren't being initialized which was causing an error. This fix involves initializing the seats, but it has to be done after the course is saved which is why it is placed here and uses a `newCourse` flag

**Comment by jsmith** [comment]
I've been testing this extensively and do not see any more bugs

**Review by tonerm43 — APPROVED** [review]
This looks good to me! I did not find any bugs while testing, and it seems to do what we want it to.

**Pull request merged**